src/java.base/share/classes/java/lang/CharacterData.java

```diff
@@ -69,7 +69,7 @@ boolean isIdeographic(int ch) {
     // Should revisit after optimization of table initialization.
 
     static final CharacterData of(int ch) {
-        if (ch >>> 8 == 0) {     // fast-path
+        if (ch >= 0 && ch <= 0xFF) {     // fast-path
```

Comment:
> `if (ch && 0xFFFFFF00 == 0) {`

This seems to perform similar to baseline:

```
Benchmark           (codePoint)  Mode  Cnt  Score   Error  Units
Characters.isDigit           48  avgt   15  0.890 ± 0.025  ns/op
Characters.isDigit         1632  avgt   15  2.174 ± 0.011  ns/op
```

Would be interesting to check the performance on non-Intel architectures. If you want to give it a spin on your M1, here's the benchmark run i used:

`make test TEST='micro:java.lang.Characters.isDigit' MICRO="OPTIONS=-p codePoint=48,1632"`

Reply:
> It seems reasonable to keep these two in sync, yes. (`CharacterData.of` could even call into `StringLatin1.canEncode`, unless that's cause for some performance anomaly)

If I update `StringLatin1.canEncode` and call into that from `CharacterData.of`, I observe no regression for the Latin1 case, but a significant regression for the non-Latin1 case. I have no idea how to explain that:

```
Benchmark           (codePoint)  Mode  Cnt  Score   Error  Units
Characters.isDigit           48  avgt   15  0.675 ± 0.029  ns/op
Characters.isDigit         1632  avgt   15  2.435 ± 0.032  ns/op
```

Reply:
Can you check what happen adding much more inputs to the dataset that includes non-latin chars as well and use `-prof perfnorm` to check what `perf` report re branches/branch-misses?

You can use SplittableRandom to pre-populate an array of inputs which sequence is "random" but still allow deterministic benchmarking and feed the benchmark method cycling the pre-computed inputs

Reply:
I created a randomized version of `Characters.isDigit` which tests with code points picked at random such that any category (Latin1, negative, different planes, unassiged) are equally probable.

Baseline:

```
Benchmark                 (codePoint)  Mode  Cnt  Score   Error  Units
Characters.isDigitRandom         1632  avgt   15  5.503 ± 0.371  ns/op
```

Current PR:

```
Benchmark                 (codePoint)  Mode  Cnt  Score   Error  Units
Characters.isDigitRandom         1632  avgt   15  5.393 ± 0.336  ns/op
```

Using StringLatin1.canEncode:

```
Benchmark                 (codePoint)  Mode  Cnt  Score   Error  Units
Characters.isDigitRandom         1632  avgt   15  5.377 ± 0.322  ns/op
```

Seems the PR still has a small improvement for this scenario. The StringLatin1.canEncode regression disappears.

In the real world ASCII/Latin1 seems to dominate most data, so this scenario is perhaps not very realistic.

I'm running this on a Mac, so cannot try `-prof perfnorm`.

Reply:
Many thanks to have tried, yep, I was curious indeed re the "StringLatin1.canEncode regression" case.
I would still modify the benchmark to use inputs (I know that will make it memory bound sadly, due to reading inputs - but the size of such inputs can be a benchmark parameter, together with the bias eg "latin","mix") "semi-randomly" generated based on the mentioned strategies/biases. 
It will benefit future tests on this, although could be provided as a separate PR.

Reply:
> The StringLatin1.canEncode regression disappears.

I mixed things up so StringLatin1.canEncode was benchmarked without the updated code.

Here are updated benchmark results:


Baseline:

```
Benchmark                 (codePoint)  Mode  Cnt  Score   Error  Units
Characters.isDigitRandom         1632  avgt   15  5.437 ± 0.235  ns/op
```

PR:

```
Benchmark                 (codePoint)  Mode  Cnt  Score   Error  Units
Characters.isDigitRandom         1632  avgt   15  5.319 ± 0.341  ns/op
````

StringLatin1.canEncode:

```
Benchmark                 (codePoint)  Mode  Cnt  Score   Error  Units
Characters.isDigitRandom         1632  avgt   15  5.447 ± 0.304  ns/op
``` 

So it seems using StringLatin1.canEncode still might have a regression also in the randomized input case.

For this PR, I suggest we update StringLatin1.canEncode to be in sync with CharacterData.of, without one calling the other. If anyone wants to investigate the regression further, than can be done outside this PR.

I have independently verified that StringLatin1.canEncode sees performance improvements using the StringIndexOf benchmark.

Reply:
We can do `Integer.compareUnsigned(ch, 0xFF) <= 0`

Reply:
Does non-short-circuit logical AND operator perform similar to baseline?  In this very specific case there is no risk of side effects due to method calls, no risk of null pointer exception because of primitives, and there is not an excessive amount of conditions.  Therefore, I would think either `&` or `&&` are viable options.

`if (ch >= 0 & ch <= 0xFF) {`

I ask just for the sake of my own learning on this topic as it relates to performance as opposed to code readability preferences.

Reply:
Both choices of operators en up with identical generated code on my Intel Mac.

For reference, I did a comparison between the generated code for the current code, and the one in this PR:

Current generated code start like this:

```
movl		%eax, -0x14000(%rsp)
pushq		%rbp
subq		$0x20, %rsp
cmpl		$1, 0x20(%r15)
jne		0x119cae072
movl		%esi, %r11d
shrl		$8, %r11d
testl		%r11d, %r11d
jne		0x119cae043
movabsq		$0x70fe03c90, %rax  ;   {oop(a 'java/lang/CharacterDataLatin1'{0x000000070fe03c90})}
``` 

While the PR generated code starts like this:

```
movl		%eax, -0x14000(%rsp)
pushq		%rbp
subq		$0x20, %rsp
cmpl		$1, 0x20(%r15)
jne		0x1171d41f2
nopw		(%rax, %rax)
cmpl		$0x100, %esi
jae		0x1171d41c5
movabsq		$0x70fe03c90, %rax  ;   {oop(a 'java/lang/CharacterDataLatin1'{0x000000070fe03c90})}
```

The rest of the code seems identical.